testing/junit/src/main/java/com/linecorp/armeria/testing/junit/server/mockwebserver/MockWebServerExtension.java

```diff
@@ -0,0 +1,174 @@
+ * {@link MockWebServerExtension#takeRequest()} to retrieve requests in the order the server received them to
+ * validate the request parameters.
+ *
+ * <p>Note, tests in a class using {@link MockWebServerExtension} cannot be run concurrently, i.e., do not set
+ * {@link org.junit.jupiter.api.parallel.Execution} to
+ * {@link org.junit.jupiter.api.parallel.ExecutionMode#CONCURRENT}.
+ *
+ * <p>For example, <pre>{@code
+ *
+ * class MyTest {
+ *
+ * {@literal @}RegisterExtension
+ * static MockWebServerExtension server = new MockWebServerExtension();
+ *
+ * {@literal @}Test
+ * void checkSomething() {
+ *     HttpClient client = HttpClient.of(server.httpUri("/"));
+ *
+ *     server.enqueue(AggregatedHttpResponse.of(HttpStatus.OK));
+ *     server.enqueue(AggregatedHttpResponse.of(HttpStatus.FORBIDDEN));
+ *
+ *     assertThat(client.get("/").aggregate().join().status()).isEqualTo(HttpStatus.OK);
+ *     assertThat(client.get("/bad").aggregate().join().status()).isEqualTo(HttpStatus.FORBIDDEN);
+ *
+ *     assertThat(server.takeRequest().path()).isEqualTo("/");
+ *     assertThat(server.takeRequest().path()).isEqualTo("/bad");
+ * }
+ *
+ * }
+ * }</pre>
+ */
+public class MockWebServerExtension extends ServerExtension implements BeforeTestExecutionCallback {
+
+    private static final Logger logger = LoggerFactory.getLogger(MockWebServerExtension.class);
+
+    private final BlockingQueue<MockResponse> mockResponses = new LinkedBlockingQueue<>();
+    private final BlockingQueue<RecordedRequest> recordedRequests = new LinkedBlockingQueue<>();
+
+    /**
+     * Enqueues the {@link MockResponse} to return to a client of this {@link MockWebServerExtension}. Multiple
+     * calls will return multiple responses in order.
+     */
+    public MockWebServerExtension enqueue(MockResponse response) {
+        mockResponses.add(response);
+        return this;
+    }
+
+    /**
+     * Enqueues the {@link AggregatedHttpResponse} to return to a client of this {@link MockWebServerExtension}.
+     * Multiple calls will return multiple responses in order.
+     */
+    public MockWebServerExtension enqueue(AggregatedHttpResponse response) {
+        mockResponses.add(MockResponse.of(response));
+        return this;
+    }
+
+    /**
+     * Returns the next {@link RecordedRequest} the server received. Call this method multiple times to retrieve
+     * the requests retrieved, in order. Will block up to 10 seconds waiting for a request.
+     */
+    @Nullable
+    public RecordedRequest takeRequest() {
+        try {
+            return recordedRequests.poll(10, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            logger.warn("Could not take request.", e);
+            Exceptions.throwUnsafely(e);
+            return null;  // Unreachable.
+        }
+    }
+
+    @Override
+    protected final void configure(ServerBuilder sb) throws Exception {
+        sb.http(0);
+        sb.https(0);
+        sb.tlsSelfSigned();
```

Comment:
Not necessarily a problem to solve in this pull request, but I'm curious if we can generate a self-signed certificate for all extensions only once.